server/src/controllers/training_controller.ts

```diff
@@ -0,0 +1,32 @@
+import express from 'express'
+import type WebSocket from 'ws'
+
+import type { Model, Task } from '@epfml/discojs'
+
+/**
+ * The Controller abstraction is commonly used in Express
+ * and comes from the MVC pattern (model-view-controller)
+ * In short, the controller is where the backend logic happens
+ * when the server receives a client request
+ * 
+ * In this case, the controller handles the training logic:
+ * what happens when a new task (DISCOllaborative) is created
+ * and what happens when receiving messages from participants
+ * of a training session.
+ * 
+ * More info on controllers:
+ * https://developer.mozilla.org/en-US/docs/Learn/Server-side/Express_Nodejs/routes
+ * 
+ */
+export abstract class TrainingController {
+
+  constructor(protected readonly task: Task) { }
+
+  abstract initTask (model: Model): void
```

Comment:
This made me realize that clients get their initial model from the server when initializing a Disco object (`Disco.fromTask` calls `client.getLatestModel`) and not when starting training (`disco.train`)
This explains why clients joining mid session still start with the initial model rather than the latest model from the session: `getLatestModel` returns the model the server has on disk but not latest model currently training

A solution would be to return the initial model when clients connect to the server within the AssignNodeID message which is used both in federated and decentralized (one less round trip compared to the current implementation in bonus)

Reply:
ho, nicely found! model origin is muddy but it's getting clearer with every PR!

> return the initial model when clients connect to the server

good idea, let's do that, that's both clearer and goes in the direction of dynamics tasks on the server :1st_place_medal:

Reply:
I'm digging into this further and we also have to take into account the browser model caching, which currently allows users to start a training session with a cached model rather than the server's base model.
This raises the question: do we allow users to start training with their local models? (which implies allowing them to continue training previously trained models) Or always overwrite it with the server's base model?

I'm leaning towards always pulling the model from the server and overwriting the local working model (as it can be saved). I don't think being able to continue training a previously trained model is important

Reply:
yeah, I'm clearly leaning in the same direction, don't really understand the need to train with a local model, let's overwrite it.